imbi/endpoints/base.py

```diff
@@ -470,3 +478,133 @@ async def patch(self, *args, **kwargs):
     @require_permission('admin')
     async def post(self, *args, **kwargs):
         await super().post(*args, **kwargs)
+
+
+@dataclasses.dataclass
+class PaginationToken:
+    """Opaque token used in "next" links by the PaginatedRequestMixin"""
+
+    start: datetime.datetime
+    limit: int
+    earliest: datetime.datetime
+
+    def with_start(self, start: datetime.datetime) -> PaginationToken:
+        """Create a new token with a different start value"""
+        return PaginationToken(start=start,
+                               limit=self.limit,
+                               earliest=self.earliest)
+
+    @classmethod
+    def from_header(cls, header_val: str) -> PaginationToken:
+        """Create a token from an opaque header value"""
+        raw_token = common.urlsafe_padded_b64decode(header_val)
+        data = umsgpack.unpackb(raw_token)
+        return PaginationToken(
+            earliest=datetime.datetime.fromisoformat(data['earliest']),
+            limit=data['limit'],
+            start=(datetime.datetime.fromisoformat(data['start']) -
+                   datetime.timedelta.resolution))
```

Comment:
I think this was supposed to be `+ datetime.timedelta.resolution`? To avoid the same kind of edge case I described below, because postgres's `BETWEEN` clause is inclusive?

Reply:
I think that the `-` is correct here since pages progress backwards in time.  This is what tries to prevent fetching the same entry twice though I still think that it could fail if there were two rows that differ only in the event time.

Reply:
Oh, yep you're right, I was thinking forwards in time.